### DISCUSSION TITLE
Detect Metadata tags that are arrays

### DISCUSSION THREAD

**Discussion by amir-titelbaum-imagenai** [Exiv2/exiv2#3029] [Q&A] [answered]
While incorporating Exiv2 into my C++ program, i came across an issue while building a JSON object from image's metadata (both Exif and XMP).

I noticed that simply using `Exiv2::Exifdatum::count()` to check whether this tag has 1 element or more, some tags (mainly "string" tags of types `asciiString`, `xmpText`  etc) report `count` as the number of characters in them.
Obviously, using `print()` or `<<value` to get the tag value as a string works perfectly. but i wished i could programatically capture the underlying types, and deliver them in a more "native" fashion.
What that means is to separate rationals to nominator and denominator, get all numbers as `Int32` or `Int64`, and most importantly: get arrays as actual `std::vector` .

**Comment by clanmills** [comment, marked as answer]
@amir-titelbaum-imagenai There is a sample application that generates JSON `exiv2json.cpp`

There isn't an API to create `std::vector`; however, you can use the dentist's method to copy the elements.  The `dentist's method` is you pull them out one at a time!

The API is documented here:
https://exiv2.org/doc/classExiv2_1_1Metadatum.html

In samples/exiv2json, I detect the type and branch appropriately.  For an array (such as rationals), I create the array, add the elements, and then store the result.  I don't remember a lot about this as I wrote it more than 10 years ago.

The most exciting and challenging part of `samples/exiv2json.cpp` was the recursion to convert the XMP structures into JSON.

```
template <class T>
void push(Jzon::Node& node,const std::string& key,T i)
{
#define ABORT_IF_I_EMTPY          \
    if (i->value().size() == 0) { \
        return;                   \
    }

    std::string value = i->value().toString();

    switch ( i->typeId() ) {
        case Exiv2::xmpText:
        ...
        case Exiv2::unsignedRational:
        case Exiv2::signedRational: {
             ABORT_IF_I_EMTPY
             Jzon::Array     arr;
             Exiv2::Rational rat = i->value().toRational();
             arr.Add(rat.first );
             arr.Add(rat.second);
             STORE(node,key,arr);
        } break;        
 ```

``` 
 $ exiv2json https://clanmills.com/Stonehenge.jpg
 {
	"Exif": {
		"Image": {
			"Make": "NIKON CORPORATION",
			"Model": "NIKON D5300",
			"Orientation": 1,
			"XResolution": [
				300,
				1
			],
			"YResolution": [
				300,
				1
			],
			"ResolutionUnit": 2,
			"Software": "Ver.1.00 ",
			"DateTime": "2015:07:16 20:25:28",
			"YCbCrPositioning": 1,
			"ExifTag": 222,
			"GPSTag": 4072
		},
		"Photo": {
			"ExposureTime": [
				10,
				4000
			],
			"FNumber": [
				100,
				10
			],
		}
        ...
	},
	"Iptc": {
		"Envelope": {
			"ModelVersion": 4,
			"CharacterSet": "%G"
		},
		"Application2": {
			"RecordVersion": 4,
			"Caption": "Classic View"
		}
	},
	"Xmp": {
		"xmp": {
			"Rating": "0",
			"ModifyDate": "2015-07-16T20:25:28+01:00"
		},
		"cm2e": {
			"Father": "Robin Mills",
			"Family": ""
		},
		"dc": {
			"description": {
				"lang": {
					"x-default": "Classic View"
				}
			},
			"Family": "Robin"
		},
		"xmlns": {
			"cm2e": "http:\/\/clanmills.com\/exiv2\/",
			"dc": "http:\/\/purl.org\/dc\/elements\/1.1\/",
			"xmp": "http:\/\/ns.adobe.com\/xap\/1.0\/"
		}
	}
}
```

**Reply by amir-titelbaum-imagenai** [reply]
Thank you @clanmills for your reply! It is much appreciated.

First i would like to point out that `samples/exiv2json.cpp` is no longer in the repo (though i did find it elsewhere on the internet).

Second, i am surprised to hear that a "sorting" mechanism is needed to pull out the values one at a time, rather than it being possible simply using the API... I am referring to the `switch case` in `exiv2json` when i say "sorting".

Is that the only way?

**Reply by clanmills** [reply]
I wouldn't use the term `sorting` as the order in which they are stored.  For example:

```
$ exiv2 -g XResolution https://clanmills.com/Stonehenge.jpg
Exif.Image.XResolution                       Rational    1  300
Exif.Thumbnail.XResolution                   Rational    1  300
$ exiv2json https://clanmills.com/Stonehenge.jpg | grep -A 3 -B 2 XResolution -
			"Model": "NIKON D5300",
			"Orientation": 1,
			"XResolution": [
				300,
				1
			],
--
		"Thumbnail": {
			"Compression": 6,
			"XResolution": [
				300,
				1
			],
$ 
```
For reasons I don't recall, I represented a RATIONAL as an array of two integers.

The library stores metadata in a list, not an array/vector. See https://exiv2.org/doc/classExiv2_1_1ExifData.html and https://exiv2.org/doc/exif_8hpp.html.  Lists and the dentist's method are close buddies.

The API offers methods such as `sortByKey()` and `sortByTag(),` presumably reordering the underlying list.

Exiv2 has used `list` to store metadata since I joined the project in 2008.  As you know, there is always more than one way to do anything. 

My final project with Exiv2 was 0.27, which shipped in December 2018, followed by numerous dot releases until June 2022.

Exiv2 0.28 was a significant project for the team to refactor/modernise the code base from C++ 1998 to C++11 and later.  I never contributed to the 0.28 code base, and the team decided to remove `samples/exiv2json.cpp` for reasons I don't know.  You can get the code from the 0.27-maintenance branch.

**Comment by MuhammadTalha121** [comment]
I'm a data scientist with C++ experience. I need to have good support for exporting JSON metadata for image processing. I've discovered that the exiv2json.cpp code has been removed in 0.28 but still exists in the 0.27-maintenance branch.
I'm going to:

-Modify the old exiv2json.cpp code to work with the current main branch
-Implement proper array type detection (so that it no longer incorrectly reports character count for string types)
-Verify that it properly supports rationals, arrays, and XMP recursion
-Implement Python bindings for direct dict/json export
-I will submit a PR in the next few days. Any advice on whether I should add this as a sample or as a feature of the library itself?

Question: Should this be a sample app or core library feature (Exiv2::toJson())? I'll start coding - expect PR in 48 hours.

**Comment by MuhammadTalha121** [comment]
I've implemented a solution in PR #3523.  
Please take a look when you have time.
